plenum/test/server/test_instance_change_provider.py

```diff
@@ -0,0 +1,203 @@
+                                            logsearch):
+    # test updating cache with view without votes
+    instance_change_db.iterator = lambda include_value=True: {
+        "3": instance_change_db_serializer.serialize(None)}.items()
+    provider = InstanceChangeProvider(tconf.OUTDATED_INSTANCE_CHANGES_CHECK_INTERVAL,
+                                      instance_change_db, time_provider)
+    assert not provider.has_view(3)
+
+    # test updating cache with Vote with incorrect timestamp format
+    instance_change_db.iterator = lambda include_value=True: {
+        "3": instance_change_db_serializer.serialize({"voter": ["a", 10.4]})}.items()
+    logs, _ = logsearch(
+        msgs=["InstanceChangeProvider: timestamp in Vote .* : .* - .* must "
+              "be of float or int type"])
+    InstanceChangeProvider(tconf.OUTDATED_INSTANCE_CHANGES_CHECK_INTERVAL,
+                           instance_change_db, time_provider)
+    assert logs
+
+    # test updating cache with Vote with incorrect reason format
+    instance_change_db.iterator = lambda include_value=True: {
+        "3": instance_change_db_serializer.serialize({"voter": [5, 10.4]})}.items()
+    logs, _ = logsearch(
+        msgs=["InstanceChangeProvider: reason in Vote .* : .* - .* must "
+              "be of int type"])
+    InstanceChangeProvider(tconf.OUTDATED_INSTANCE_CHANGES_CHECK_INTERVAL,
+                           instance_change_db, time_provider)
+    assert logs
+
+    # test updating cache with incorrect view_no format
+    instance_change_db.iterator = lambda include_value=True: {
+        "a": instance_change_db_serializer.serialize({"voter": [5, 25]})}.items()
+    logs, _ = logsearch(
+        msgs=["InstanceChangeProvider: view_no='.*' "
+              "must be of int type"])
+    InstanceChangeProvider(tconf.OUTDATED_INSTANCE_CHANGES_CHECK_INTERVAL,
+                           instance_change_db, time_provider)
+    assert logs
+
+
+def test_remove_view(instance_change_provider):
+    frm = "Node1"
+    view_no = 2
+
+    instance_change_provider.add_vote(InstanceChange(view_no - 1,
+                                                     Suspicions.PRIMARY_DEGRADED.code), frm)
+    instance_change_provider.add_vote(InstanceChange(view_no,
+                                                     Suspicions.PRIMARY_DEGRADED.code), frm)
+
+    assert instance_change_provider.has_view(view_no - 1)
+    assert instance_change_provider.has_view(view_no)
+    assert instance_change_provider.has_inst_chng_from(view_no - 1, frm)
+    assert instance_change_provider.has_inst_chng_from(view_no, frm)
+
+    instance_change_provider.remove_view(view_no)
+
+    assert not instance_change_provider.has_view(view_no - 1)
+    assert not instance_change_provider.has_view(view_no)
+    assert not instance_change_provider.has_inst_chng_from(view_no - 1, frm)
+    assert not instance_change_provider.has_inst_chng_from(view_no, frm)
```

Comment:
It would be also nice to check that record about view is actually deleted from database, so we don't have small, but potentially unbounded growth.

Reply:
Added. Thanks.